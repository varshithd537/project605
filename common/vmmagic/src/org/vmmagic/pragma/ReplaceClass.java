/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Eclipse Public License (EPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/eclipse-1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package org.vmmagic.pragma;

import java.lang.annotation.Retention;
import java.lang.annotation.Target;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.ElementType;
import org.vmmagic.Pragma;

/**
 * The ReplaceClass annotation indicates that the members of this class annotated with
 * <code>ReplaceMember</code> will replace members who have same name and same descriptor
 * in the targeting class. The <code>value()</code> of this annotation should be the
 * descriptor of targeting class.
 */
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.TYPE)
@Pragma
public @interface ReplaceClass {

  /**
   * @return the name of the class to be replaced, e.g. {@code java.lang.System}
   */
  String className();
}
