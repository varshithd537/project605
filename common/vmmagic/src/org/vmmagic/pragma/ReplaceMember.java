/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Eclipse Public License (EPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/eclipse-1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package org.vmmagic.pragma;

import java.lang.annotation.Retention;
import java.lang.annotation.Target;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.ElementType;
import org.vmmagic.Pragma;

/**
 * The ReplaceMember annotation indicates that this member will replace the member
 * who has the same name, or has the name given by {@code value()}, and same
 * descriptor in the targeting class which is referred by the {@code value()} of
 * {@code ReplaceClass} annotation.
 */
@Retention(RetentionPolicy.RUNTIME)
@Target({ElementType.FIELD, ElementType.METHOD})
@Pragma
public @interface ReplaceMember {
  String value() default "";
  String descriptor() default "";
}
