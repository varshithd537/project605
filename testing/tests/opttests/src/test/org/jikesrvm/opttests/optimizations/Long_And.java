/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Eclipse Public License (EPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/eclipse-1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package test.org.jikesrvm.opttests.optimizations;

import org.vmmagic.pragma.NoInline;

public class Long_And {

  private static final long TwoToThePowerOf32 = 1L << 32;

  public static void main(String[] args) {
    Long_And long_and = new Long_And();
    long_and.main();
  }

  private void main() {
    System.out.println("--- Tests for long and ---");
    System.out.println("--- Clear everything ---");
    clearEverything(0L);
    clearEverything(-1L);
    clearEverything(1L);
    clearEverything(Long.MAX_VALUE);
    clearEverything(Long.MIN_VALUE);
    clearEverything(Integer.MAX_VALUE);
    clearEverything(Integer.MIN_VALUE);
    System.out.println("--- Don't clear anything ---");
    dontClearAnything(0L);
    dontClearAnything(-1L);
    dontClearAnything(1L);
    dontClearAnything(Long.MAX_VALUE);
    dontClearAnything(Long.MIN_VALUE);
    dontClearAnything(Integer.MAX_VALUE);
    dontClearAnything(Integer.MIN_VALUE);
    System.out.println("--- And with self ---");
    andWithSelf(0L);
    andWithSelf(-1L);
    andWithSelf(1L);
    andWithSelf(Long.MAX_VALUE);
    andWithSelf(Long.MIN_VALUE);
    andWithSelf(Integer.MAX_VALUE);
    andWithSelf(Integer.MIN_VALUE);
    System.out.println("--- Reduce to 1 bit ---");
    reduceTo1Bit(0L);
    reduceTo1Bit(-1L);
    reduceTo1Bit(1L);
    reduceTo1Bit(Long.MAX_VALUE);
    reduceTo1Bit(Long.MIN_VALUE);
    reduceTo1Bit(Integer.MAX_VALUE);
    reduceTo1Bit(Integer.MIN_VALUE);
    reduceTo1Bit(2L);
    System.out.println("--- Clear upper 32 ---");
    clearUpper32(0L);
    clearUpper32(-1L);
    clearUpper32(1L);
    clearUpper32(Long.MAX_VALUE);
    clearUpper32(Long.MIN_VALUE);
    clearUpper32(Integer.MAX_VALUE);
    clearUpper32(Integer.MIN_VALUE);
    clearUpper32(TwoToThePowerOf32);
    System.out.println("--- Clear lower 32 ---");
    clearLower32(0L);
    clearLower32(-1L);
    clearLower32(1L);
    clearLower32(Long.MAX_VALUE);
    clearLower32(Long.MIN_VALUE);
    clearLower32(Integer.MAX_VALUE);
    clearLower32(Integer.MIN_VALUE);
    clearLower32(TwoToThePowerOf32);
    System.out.println("--- And with 2^32 ---");
    andWith2ToThePowerOf32(0L);
    andWith2ToThePowerOf32(-1L);
    andWith2ToThePowerOf32(1L);
    andWith2ToThePowerOf32(Long.MAX_VALUE);
    andWith2ToThePowerOf32(Long.MIN_VALUE);
    andWith2ToThePowerOf32(Integer.MAX_VALUE);
    andWith2ToThePowerOf32(Integer.MIN_VALUE);
    andWith2ToThePowerOf32(TwoToThePowerOf32);
  }

  @NoInline
  private static void clearEverything(long val) {
    long res = val & 0L;
    System.out.println(Long.toString(val) + " & 0L: " + Long.toString(res));
  }

  @NoInline
  private static void dontClearAnything(long val) {
    long res = val & -1L;
    System.out.println(Long.toString(val) + " & -1L: " + Long.toString(res));
  }

  @NoInline
  private static void andWithSelf(long val) {
    long res = val & val;
    System.out.println(Long.toString(val) + " & " + Long.toString(val) + "L: " + Long.toString(res));
  }

  @NoInline
  private static void reduceTo1Bit(long val) {
    long res = val & 1L;
    System.out.println(Long.toString(val) + " & 1L: " + Long.toString(res));
  }

  @NoInline
  private static void clearUpper32(long val) {
    long res = val & 0x00000000FFFFFFFFL;
    System.out.println(Long.toString(val) + " & 0x00000000FFFFFFFFL: " + Long.toString(res));
  }

  @NoInline
  private static void clearLower32(long val) {
    long res = val & 0xFFFFFFFF00000000L;
    System.out.println(Long.toString(val) + " & 0xFFFFFFFF00000000L: " + Long.toString(res));
  }

  @NoInline
  private static void andWith2ToThePowerOf32(long val) {
    long res = val & 0x00000001000000000L;
    System.out.println(Long.toString(val) + " & 0x00000001000000000L: " + Long.toString(res));
  }

}
