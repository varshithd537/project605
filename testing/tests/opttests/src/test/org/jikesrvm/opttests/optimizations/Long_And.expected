--- Tests for long and ---
--- Clear everything ---
0 & 0L: 0
-1 & 0L: 0
1 & 0L: 0
9223372036854775807 & 0L: 0
-9223372036854775808 & 0L: 0
2147483647 & 0L: 0
-2147483648 & 0L: 0
--- Don't clear anything ---
0 & -1L: 0
-1 & -1L: -1
1 & -1L: 1
9223372036854775807 & -1L: 9223372036854775807
-9223372036854775808 & -1L: -9223372036854775808
2147483647 & -1L: 2147483647
-2147483648 & -1L: -2147483648
--- And with self ---
0 & 0L: 0
-1 & -1L: -1
1 & 1L: 1
9223372036854775807 & 9223372036854775807L: 9223372036854775807
-9223372036854775808 & -9223372036854775808L: -9223372036854775808
2147483647 & 2147483647L: 2147483647
-2147483648 & -2147483648L: -2147483648
--- Reduce to 1 bit ---
0 & 1L: 0
-1 & 1L: 1
1 & 1L: 1
9223372036854775807 & 1L: 1
-9223372036854775808 & 1L: 0
2147483647 & 1L: 1
-2147483648 & 1L: 0
2 & 1L: 0
--- Clear upper 32 ---
0 & 0x00000000FFFFFFFFL: 0
-1 & 0x00000000FFFFFFFFL: 4294967295
1 & 0x00000000FFFFFFFFL: 1
9223372036854775807 & 0x00000000FFFFFFFFL: 4294967295
-9223372036854775808 & 0x00000000FFFFFFFFL: 0
2147483647 & 0x00000000FFFFFFFFL: 2147483647
-2147483648 & 0x00000000FFFFFFFFL: 2147483648
4294967296 & 0x00000000FFFFFFFFL: 0
--- Clear lower 32 ---
0 & 0xFFFFFFFF00000000L: 0
-1 & 0xFFFFFFFF00000000L: -4294967296
1 & 0xFFFFFFFF00000000L: 0
9223372036854775807 & 0xFFFFFFFF00000000L: 9223372032559808512
-9223372036854775808 & 0xFFFFFFFF00000000L: -9223372036854775808
2147483647 & 0xFFFFFFFF00000000L: 0
-2147483648 & 0xFFFFFFFF00000000L: -4294967296
4294967296 & 0xFFFFFFFF00000000L: 4294967296
--- And with 2^32 ---
0 & 0x00000001000000000L: 0
-1 & 0x00000001000000000L: 68719476736
1 & 0x00000001000000000L: 0
9223372036854775807 & 0x00000001000000000L: 68719476736
-9223372036854775808 & 0x00000001000000000L: 0
2147483647 & 0x00000001000000000L: 0
-2147483648 & 0x00000001000000000L: 68719476736
4294967296 & 0x00000001000000000L: 0
